### PR TITLE
add enable browser detection and some tests

### LINKED ISSUE #18448: Add an option to disable fingerprinting / config_id entirely
## Update - the specs

- Fix in 5.0.0 - See the proposed solution in https://github.com/matomo-org/matomo/issues/18448#issuecomment-1512338767
- Ideally we would backport but it might be OK in that case not to backport

## Summary

This is continuation of #16361 and the [related discussion](https://forum.matomo.org/t/does-device-fingerprinting-require-consent/42869) in the forums. Right now, the stance of Matomo is that in cookie-less mode, [it doesn't use fingerprinting](https://matomo.org/faq/general/how-is-the-visitor-config_id-processed/) so it doesn't require consent. Under GDPR this line of thought is perfectly sound and reasonable. However, under the ePrivacy directive (which is now implemented in German law since 1 December, 2021 with the [TTDSG](https://dsgvo-gesetz.de/ttdsg/)), this is once again a debatable point.

The key phrase in the ePrivacy directive is the 'access to information' already stored on the user's device. Of course, it's not entirely clear what this means in practice. But call it a fingerprint or not, Matomo does [access some data from the user's device](https://github.com/matomo-org/matomo/blob/4.x-dev/js/piwik.js#L3130-L3187) to create the config_id (in particular, the screen size and supported mime types). It's really up for debate if this constitutes 'access to information' – but as long as there's no legal precedence, some clients will want to cover all their bases. So it will once again be necessary to require consent before using Matomo, even without cookies or 'real' fingerprinting in place. So having an option to disable the device detection further would be great. Instead, the config_id could be based solely on User-Agent and anonymised IP address (which the server receives implicitly, so no data access is required).

Of course, this would further limit the usefulnes of some reports. But this would be acceptable in order to be 100% sure that using Matomo without consent is completely safe, legally speaking.

### PR DESCRIPTION
**Pull request by peterhashair**
### Description:

Fixes: #18448 

add enable browser detection and some tests

[develop document](https://github.com/matomo-org/developer-documentation/pull/606)

[FAQ draft](https://matomo.org/wp-admin/post.php?post=48524&action=edit&lang=en)

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

**Comment by peterhashair** [comment]
build js

**Comment by peterhashair** [comment]
@sgiehl sorry to bother, just checking what would be the best approach for piwik.js _pag.push tests I found here, https://github.com/matomo-org/plugin-AnonymousPiwikUsageMeasurement/pull/68 add it here. not sure if that's correct.

**Review by sgiehl — CHANGES_REQUESTED** [review]
@peterhashair seems there are some JSLint errors, that need to be fixed. Otherwise looks good.
@tsteur would that be something that should be included in 4.7.0 release?

**Comment by tsteur** [comment]
Very low risk of breaking something so that could be included.

Be also good to add more tests. We could make sure no browser resolution and detected plugins information is in the tracking request when disabled. Seems the other usages of `detectBrowserFeatures` might be less testable because they are partially including random uids etc.

We'd also need to update the previously FAQ to mention the new method.

We also need to update https://developer.matomo.org/api-reference/tracking-javascript to mention the new methods.

BTW are these new methods mentioned in the developer changelog?

This is documented in https://developer.matomo.org/guides/jstracker-core#adding-a-new-public-js-tracker-method

**Comment by peterhashair** [comment]
@tsteur regarding the tests, I was thinking this test include the browser resolution check. Should I do a fake tracking in Integration tests? 
```javascript 
    test("Browser detector feature Disable and enable", function() {
      var pattern = /res/;
      var tracker = Piwik.getTracker();
      tracker.disableBrowserFeatureDetection();
      var request = tracker.getRequest('hello=world');
      equal(pattern.test(request), false, 'Disable browser fingerprint');

      tracker.enableBrowserFeatureDetection();
      var request = tracker.getRequest('hello=world');
      equal(pattern.test(request), true, 'Enable browser fingerprint set');

    });
```

**Comment by tsteur** [comment]
Above test looks good. You'll also want to check this for the other tracking parameters. And could also adjust the messages like `Disable browser fingerprint'` change it to `When browser fingerprint is disabled the request should not include browser resolution'`

**Comment by sgiehl** [comment]
@peterhashair There are still some JSLint errors to be fixed. If you need help with that, let me know.

**Comment by peterhashair** [comment]
@sgiehl found it, fixed it. But it seems like the release check failed because the file size increased?

File: tests/javascript/index.php [hunk 1]
```diff
@@ -5220,6 +5220,27 @@ function testWithWrappedConsole(callback) {
...
+    test("Browser detector feature Disable and enable", function() {
+        var pattern = /(res=)|(cookie=)/;
+        var tracker = Piwik.getTracker();
+        var siteIdPattern = /idsite/;
+
+        tracker.enableBrowserFeatureDetection();
+        var requestWithFingerprint = tracker.getRequest('hello=world');
+
+        equal(siteIdPattern.test(requestWithFingerprint), true);
+        equal(pattern.test(requestWithFingerprint), true, 'When browser fingerprint is enabled the request should include browser resolution and cookie');
```
Comment (sgiehl) [inline tests/javascript/index.php]: Doesn't the pattern actually tests if the string contains resolution or cookie?

File: tests/javascript/index.php [hunk 2]
```diff
@@ -5220,6 +5220,27 @@ function testWithWrappedConsole(callback) {
...
+
+        tracker.disableBrowserFeatureDetection();
+        var requestWithoutFingerprint = tracker.getRequest('hello=world');
+
+        equal(siteIdPattern.test(requestWithoutFingerprint), true);
+        equal(pattern.test(requestWithoutFingerprint), false, 'When browser fingerprint is disabled the request should not include browser resolution or cookie');
+
+
+
```
Comment (sgiehl) [inline tests/javascript/index.php]: maybe try to avoid too many useless empty lines

**Review by sgiehl — CHANGES_REQUESTED** [review]
Left some minor comments. Otherwise looks good.

File: tests/javascript/index.php (hunk 1 above)
Comment (peterhashair) [reply tests/javascript/index.php]: @sgiehl ah should I change this part ` var pattern = /(res=)|(cookie=)/;`

**Pull request merged**